### PR TITLE
refactor(ecmascript): remove unwrap_module ast cloning

### PR DESCRIPTION
**Pull request by kwonoj**
### Description

Closing WEB-1024 on turbopack end.

### CODE REVIEW

File: crates/turbopack-ecmascript/src/transform/mod.rs [hunk 1]
```diff
@@ -363,21 +363,6 @@ pub fn remove_shebang(program: &mut Program) {
...
-        Program::Module(module) => Program::Module(module.take()),
-        Program::Script(s) => Program::Module(Module {
-            span: s.span,
-            body: s
-                .body
-                .iter()
-                .map(|stmt| ModuleItem::Stmt(stmt.clone()))
-                .collect(),
-            shebang: s.shebang.clone(),
-        }),
```
Comment (sokra) [inline crates/turbopack-ecmascript/src/transform/mod.rs:377]: Just noticing that the behavior is not the same. `unwrap_module_program` converts `script` to `module`. We need to keep that behavior

**Review by sokra — CHANGES_REQUESTED** [review]

File: crates/turbopack-ecmascript/src/transform/mod.rs (hunk 1 above)
Comment (kwonoj) [reply crates/turbopack-ecmascript/src/transform/mod.rs:377]: Ok, converted to use visitMut instead: 

https://github.com/vercel/turbo/pull/4884/files#diff-b1b0731c098d856a121581436124b8fd726411b872ed875aacce400c825b0fbfR375

**Review by jridgewell — APPROVED** [review]

File: crates/turbopack-ecmascript/src/transform/mod.rs [hunk 2]
```diff
@@ -221,7 +221,11 @@ impl EcmascriptInputTransform {
                     ..Default::default()
                 };
 
-                let module_program = unwrap_module_program(program);
+                let mut module_program =
+                    std::mem::replace(program, Program::Module(Module::dummy()));
+
+                let mut unwrap_visitor = UnwrapModuleProgram;
+                module_program.visit_mut_with(&mut unwrap_visitor);
```
Comment (sokra) [inline crates/turbopack-ecmascript/src/transform/mod.rs:228]:
```suggestion
                let mut program =
                    std::mem::replace(program, Program::Module(Module::dummy()));

                let module_program = if let Program::Script(s) = program {
                    Program::Module(Module {
                        span: s.span,
                        body: s
                            .body
                            .iter()
                            .map(|stmt| ModuleItem::Stmt(stmt.clone()))
                            .collect(),
                        shebang: s.shebang.clone(),
                    })
                } else {
                    program
                }
```

File: crates/turbopack-ecmascript/src/transform/mod.rs [hunk 3]
```diff
@@ -355,18 +359,23 @@ pub fn remove_shebang(program: &mut Program) {
...
+
+    // If given program is a script, replace it with a module.
+    fn visit_mut_program(&mut self, program: &mut Program) {
+        if let Program::Script(s) = program {
+            *program = Program::Module(Module {
+                span: s.span,
+                body: s
+                    .body
+                    .iter()
+                    .map(|stmt| ModuleItem::Stmt(stmt.clone()))
```
Comment (sokra) [inline crates/turbopack-ecmascript/src/transform/mod.rs:374]: It's still cloning...

**Review by sokra — CHANGES_REQUESTED** [review]
You don't need a Visitor

**Review by sokra — APPROVED** [review]

**Pull request merged**